### PR TITLE
enhance video streaming performance for android

### PR DESCRIPTION
**Pull request by BrandonHe**
This proposal aims to enhance SDL Android library for massive data transmission performance for video streaming and projection by refactor default I/O pipe write/read class.

### CODE REVIEW

File: proposals/nnnn-enhance-video-streaming-performance-for-android.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+* Status: **Awaiting review**
+* Impacted Platforms: [Android]
+
+## Introduction
+
+This proposal aims to enhance SDL Android library for massive data transmission performance for video streaming and projection by refactor default I/O pipe write/read class.
+
+## Motivation
+
+The SDL Android library are currently using default java.io package, it uses `PipedInputStream` via `write()` function to stream data **Byte by Byte**, because CPU operation is very fast, in most cases, the operation performance as good as we expected. But for video streaming and projection, it is a botterneck for transmission according to test. It will cause very frequently CPU resources switch between read thread and write thread. Thus has an effect on video streaming performance. Another motivation here is that if read thread get the critical resource to read after write thread just write one byte or serveral bytes, this leads to slice one frame data to send by many times. It lso reduces the efficiency of data transmission and increase the CPU workload.
```
Comment (theresalech) [inline proposals/nnnn-enhance-video-streaming-performance-for-android.md]: Please make the following changes in this paragraph:

- Change "performance" to "performs"
- Change "botterneck" to "bottleneck" 
- Change "frequently" to "frequent"
- Change "lso" to "also"

File: proposals/nnnn-enhance-video-streaming-performance-for-android.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+## Proposed solution
+
+The solution in this proposal is to refactor the class `PipedInputStream` and  the class `PipedOutputStream`, it gives parameter byteOffset and parameter byteCount in recevie function and write function. These will help these two function to call `System.arraycopy()` to receive and write block data into buffer one time.
```
Comment (theresalech) [inline proposals/nnnn-enhance-video-streaming-performance-for-android.md]: Please change "recevie" to "receive" and "two function" to "two functions"

File: proposals/nnnn-enhance-video-streaming-performance-for-android.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+```
+
+## Potential downsides
+
+N/A
+
+## Impact on existing code
+
+There is no impact for the existing code. The changes in this proposal is to remove the default io PipedInputStream and PipedOutputStream packge and reimport PipedInputStream and PipedOutputStream from component streaming. The developer would not need to change their previous code if they do not wish to use the new library for video streaming.
```
Comment (theresalech) [inline proposals/nnnn-enhance-video-streaming-performance-for-android.md]: Please update to "changes in this proposal are to remove" and "packge" to "package"

File: proposals/nnnn-enhance-video-streaming-performance-for-android.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+ 
+## Alternatives considered
+
+There is a optional solution to improve the video streaming performance is to straightly send data from encoder byte buffer and handler to avoid using the pipe stream.
```
Comment (theresalech) [inline proposals/nnnn-enhance-video-streaming-performance-for-android.md]: Please update to "There is an optional solution to improve the video streaming performance by directly sending data..."

**Comment by joeygrover** [comment]
Please change the description of this pull request to the introduction of the proposal as stated in the [pull request template](https://github.com/smartdevicelink/sdl_evolution/blob/revert-182-0058-video-streaming-capabilities.md/.github/PULL_REQUEST_TEMPLATE.md).

**Comment by BrandonHe** [comment]
@theresalech 
Thanks for your comments, I have fixed the spelling/grammatical error.
there is no other dependency in this PR.

**Comment by BrandonHe** [comment]
@joeygrover 
Thanks for your comment, I have changed the initial text of pull request to the _exact_ text of the introduction of my proposal

**Pull request merged**